src/pe_reports/metrics.py

```diff
@@ -12,25 +22,571 @@ def __init__(self, start_date, end_date, org_uid):
+                columns=[
+                    "Domain",
+                    "IPv4",
+                    "IPv6",
+                    "Mail Server",
+                    "Name Server",
+                ]
+            )
+        return domain_sum
+
+    def utlds(self):
+        """Return count of unique top level domains."""
+        mal_df = self.df_mal
+
+        if mal_df > 0:
+            mal_df["tld"] = (
+                mal_df["domain_permutation"]
+                .str.split(".")
+                .str[-1]
+                .str.split("/")
+                .str[0]
+            )
+            utlds = len(mal_df["tld"].unique())
+        else:
+            utlds = 0
+
+        return utlds
+
+
+class Malware_Vulns:
+    """Malware and Vulnerabilities Class."""
+
+    def __init__(self, start_date, end_date, org_uid):
+        """Initialize Shodan vulns and malware class."""
+        self.start_date = start_date
+        self.end_date = end_date
+        self.org_uid = org_uid
+        insecure_df = query_shodan(
+            org_uid,
+            start_date,
+            end_date,
+            "shodan_insecure_protocols_unverified_vulns",
+        )
+        self.insecure_df = insecure_df
+
+        vulns_df = query_shodan(org_uid, start_date, end_date, "shodan_verified_vulns")
+        vulns_df["port"] = vulns_df["port"].astype(str)
+        self.vulns_df = vulns_df
+
+        assets_df = query_shodan(org_uid, start_date, end_date, "shodan_assets")
+        self.assets_df = assets_df
+
+    def protocol_count(self):
+        """Return a count for each insecure protocol."""
+        insecure_df = self.insecure_df
+        insecure = insecure_df[insecure_df["type"] == "Insecure Protocol"]
+        insecure = insecure[
+            (insecure["protocol"] != "http") & (insecure["protocol"] != "smtp")
+        ]
+        risky_assets = insecure[["ip", "protocol"]].drop_duplicates(keep="first")
+
+        # Horizontal bar: insecure protocol count
+        pro_count = risky_assets.groupby(["protocol"], as_index=False)["protocol"].agg(
+            {"id_count": "count"}
+        )
+        return pro_count
+
+    def risky_assets(self):
+        """Return dataframe of ip and risky protocols."""
+        insecure_df = self.insecure_df
+        insecure = insecure_df[insecure_df["type"] == "Insecure Protocol"]
+        insecure = insecure[
+            (insecure["protocol"] != "http") & (insecure["protocol"] != "smtp")
+        ]
+        risky_assets = insecure[["ip", "protocol"]].drop_duplicates(keep="first")
+        return risky_assets
+
+    def risky_ports_count(self):
+        """Return total count of insecure protocols."""
+        insecure_df = self.insecure_df
+        insecure = insecure_df[insecure_df["type"] == "Insecure Protocol"]
+        insecure = insecure[
+            (insecure["protocol"] != "http") & (insecure["protocol"] != "smtp")
+        ]
+        risky_assets = insecure[["ip", "protocol"]].drop_duplicates(keep="first")
```

Comment:
You've got a lot of repeated code in these three functions, namely:
```
insecure_df = self.insecure_df
insecure = insecure_df[insecure_df["type"] == "Insecure Protocol"]
insecure = insecure[
    (insecure["protocol"] != "http") & (insecure["protocol"] != "smtp")
]
risky_assets = insecure[["ip", "protocol"]].drop_duplicates(keep="first")
```
You should consider moving that code to a separate function.

Reply:
@dav3r Addressed this comment in the following commit: [051e84b]( https://github.com/cisagov/pe-reports/pull/99/commits/051e84b856e0c6bf33b31d24fcdaf704ac199406l)